File: iOS/ssabree/ssabree/Features/Join/JoinViewModel.swift
import SwiftUI
import Observation

// MARK: - Debug Log Helper
#if DEBUG
func joinLog(_ message: String) {
    print("[JoinVM] \(message)")
}
#else
func joinLog(_ message: String) {}
#endif

// MARK: - Join Error

enum JoinError: LocalizedError {
    case validation(String)
    case server(String)
    case network
    case unknown

    var errorDescription: String? {
        switch self {
        case .validation(let msg): return msg
        case .server(let msg): return msg
        case .network: return "네트워크 연결을 확인해주세요"
        case .unknown: return "알 수 없는 오류가 발생했습니다"
        }
    }
}

// MARK: - Join UI State
// Android JoinUiState와 동일한 구조

struct JoinUiState {
    // Step 1: 개인 정보
    var name: String = ""
    var studentId: String = ""

    // 기수 선택
    var generations: [Int] = [14, 15]
    var selectedGeneration: Int? = nil
    var isGenerationDropdownExpanded: Bool = false

    // 캠퍼스 선택
    var campuses: [Campus] = []
    var selectedCampus: Campus? = nil
    var isCampusDropdownExpanded: Bool = false
    var isLoadingCampuses: Bool = false

    // 반 선택 - Android와 동일하게 Ban 객체 사용
    var campusClasses: [Ban] = []  // 캠퍼스의 전체 반 목록
    var classes: [Ban] = []         // 기수로 필터링된 반 목록
    var selectedClass: Ban? = nil
    var isClassDropdownExpanded: Bool = false
    var isLoadingClasses: Bool = false

    // Step 2: Mattermost 인증
    var mattermostId: String = ""
    var verificationCode: String = ""
    var isCodeSent: Bool = false
    var isVerified: Bool = false
    var isSendingCode: Bool = false
    var isVerifying: Bool = false

    // Step 3: 계정 정보
    var email: String = ""
    var isCheckingEmail: Bool = false
    var isEmailAvailable: Bool? = nil
    var password: String = ""
    var passwordConfirm: String = ""

    // 공통
    var step: Int = 1
    var isLoading: Bool = false
    var error: JoinError? = nil
    var isJoinSuccess: Bool = false

    // Computed properties - Android와 동일
    var isStep1Valid: Bool {
        !name.isEmpty &&
        !studentId.isEmpty &&
        selectedGeneration != nil &&
        selectedCampus != nil &&
        selectedClass != nil
    }

    var isStep2Valid: Bool {
        verificationCode.count == 6
    }

    var isStep3Valid: Bool {
        !email.isEmpty &&
        !password.isEmpty &&
        !passwordConfirm.isEmpty &&
        password == passwordConfirm &&
        !isLoading
    }

    var passwordMismatch: Bool {
        !passwordConfirm.isEmpty && password != passwordConfirm
    }
}

// MARK: - Join ViewModel

@MainActor
@Observable
final class JoinViewModel {
    var uiState = JoinUiState()

    private let authRepository: AuthRepository
    private let campusRepository: CampusRepository

    init(authRepository: AuthRepository, campusRepository: CampusRepository) {
        self.authRepository = authRepository
        self.campusRepository = campusRepository
        loadCampuses()
    }

    // MARK: - Load Data

    private func loadCampuses() {
        Task {
            joinLog("loadCampuses() called")
            uiState.isLoadingCampuses = true

            let result = await campusRepository.getCampuses()

            switch result {
            case .success(let campusList):
                joinLog("loadCampuses SUCCESS: \(campusList.count) campuses - \(campusList.map { $0.name })")
                uiState.isLoadingCampuses = false
                uiState.campuses = campusList
            case .failure(let error):
                joinLog("loadCampuses FAILED: \(error)")
                uiState.isLoadingCampuses = false
                uiState.error = toJoinError(error)
            }
        }
    }

    private func loadClasses(campusId: Int) {
        Task {
            joinLog("loadClasses(campusId: \(campusId)) called")
            uiState.isLoadingClasses = true
            uiState.classes = []
            uiState.selectedClass = nil

            let result = await campusRepository.getClasses(campusId: campusId)

            switch result {
            case .success(let classList):
                joinLog("loadClasses SUCCESS: \(classList.count) classes")
                joinLog("classList generations: \(Set(classList.compactMap { $0.generation }))")
                joinLog("classList campusIds: \(Set(classList.map { $0.campusId }))")
                uiState.isLoadingClasses = false
                uiState.campusClasses = classList
                // 기수와 캠퍼스로 필터링
                let filtered = filterClasses(source: classList, generation: uiState.selectedGeneration, campusId: uiState.selectedCampus?.id)
                joinLog("filtered for gen \(uiState.selectedGeneration ?? -1), campus \(uiState.selectedCampus?.id ?? -1): \(filtered.count) classes")
                uiState.classes = filtered
                uiState.selectedClass = nil
            case .failure(let error):
                joinLog("loadClasses FAILED: \(error)")
                uiState.isLoadingClasses = false
                uiState.error = toJoinError(error)
            }
        }
    }

    // MARK: - Filter Classes by Generation and Campus
    // Android: filterClasses(source: List<Ban>, generation: Int?)

    private func filterClasses(source: [Ban], generation: Int?, campusId: Int? = nil) -> [Ban] {
        guard let generation = generation else { return [] }
        var filtered = source.filter { $0.generation == generation }
        // 캠퍼스 ID로도 필터링 (API가 전체 반을 반환하는 경우)
        if let campusId = campusId {
            filtered = filtered.filter { $0.resolvedCampusId == campusId }
        }
        return filtered
    }

    // MARK: - Step Navigation

    func goToStep(_ step: Int) {
        uiState.step = step
        uiState.error = nil
    }

    func goBack() -> Bool {
        switch uiState.step {
        case 1:
            return false
        case 2:
            uiState.step = 1
            uiState.error = nil
            uiState.mattermostId = ""
            uiState.verificationCode = ""
            uiState.isCodeSent = false
            uiState.isVerified = false
            uiState.isSendingCode = false
            uiState.isVerifying = false
            uiState.isGenerationDropdownExpanded = false
            return true
        case 3:
            uiState.step = 2
            uiState.error = nil
            uiState.mattermostId = ""
            uiState.verificationCode = ""
            uiState.isCodeSent = false
            uiState.isVerified = false
            uiState.isSendingCode = false
            uiState.isVerifying = false
            uiState.isGenerationDropdownExpanded = false
            uiState.email = ""
            uiState.isCheckingEmail = false
            uiState.isEmailAvailable = nil
            uiState.password = ""
            uiState.passwordConfirm = ""
            uiState.isLoading = false
            return true
        default:
            return false
        }
    }

    // MARK: - Step 1: Personal Info

    func onNameChange(_ value: String) {
        uiState.name = value
    }

    func onStudentIdChange(_ value: String) {
        uiState.studentId = value.filter { $0.isNumber }
    }

    // MARK: - 기수 선택 (Generation)

    func onGenerationDropdownClick() {
        joinLog("onGenerationDropdownClick()")
        uiState.isGenerationDropdownExpanded = true
    }

    func onGenerationDropdownDismiss() {
        joinLog("onGenerationDropdownDismiss()")
        uiState.isGenerationDropdownExpanded = false
    }

    func onGenerationSelected(_ generation: Int) {
        joinLog("onGenerationSelected(\(generation))")
        joinLog("campusClasses count: \(uiState.campusClasses.count)")
        // Android와 동일: 기수 선택 시 campusClasses에서 필터링
        let filteredClasses = filterClasses(source: uiState.campusClasses, generation: generation, campusId: uiState.selectedCampus?.id)
        joinLog("filteredClasses count: \(filteredClasses.count)")
        uiState.selectedGeneration = generation
        uiState.isGenerationDropdownExpanded = false
        uiState.classes = filteredClasses
        uiState.selectedClass = nil
        joinLog("after onGenerationSelected - selectedGeneration: \(uiState.selectedGeneration ?? -1), campus: \(uiState.selectedCampus?.name ?? "nil"), classes: \(uiState.classes.count)")
    }

    // MARK: - 캠퍼스 선택 (Campus)

    func onCampusDropdownClick() {
        joinLog("onCampusDropdownClick() - campuses: \(uiState.campuses.count)")
        uiState.isCampusDropdownExpanded = true
    }

    func onCampusDropdownDismiss() {
        joinLog("onCampusDropdownDismiss()")
        uiState.isCampusDropdownExpanded = false
    }

    func onCampusSelected(_ campus: Campus) {
        joinLog("onCampusSelected(\(campus.name))")
        uiState.selectedCampus = campus
        uiState.isCampusDropdownExpanded = false
        uiState.selectedClass = nil
        uiState.classes = []
        uiState.campusClasses = []
        loadClasses(campusId: campus.id)
    }

    // MARK: - 반 선택 (Class)

    func onClassDropdownClick() {
        joinLog("onClassDropdownClick() - campus: \(uiState.selectedCampus?.name ?? "nil"), gen: \(uiState.selectedGeneration ?? -1), classes: \(uiState.classes.count)")
        // Android와 동일: 캠퍼스와 기수가 선택되어야 클릭 가능
        if uiState.selectedCampus != nil && uiState.selectedGeneration != nil {
            uiState.isClassDropdownExpanded = true
        }
    }

    func onClassDropdownDismiss() {
        joinLog("onClassDropdownDismiss()")
        uiState.isClassDropdownExpanded = false
    }

    func onClassSelected(_ ban: Ban) {
        joinLog("onClassSelected(\(ban.name))")
        uiState.selectedClass = ban
        uiState.isClassDropdownExpanded = false
    }

    // MARK: - Step 2: Mattermost Verification

    func onMattermostIdChange(_ value: String) {
        uiState.mattermostId = value
        uiState.error = nil
    }

    func onVerificationCodeChange(_ value: String) {
        uiState.verificationCode = String(value.filter { $0.isNumber }.prefix(6))
        uiState.error = nil
    }

    func sendVerificationCode() {
        let mmId = uiState.mattermostId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let generation = uiState.selectedGeneration, !mmId.isEmpty else {
            joinLog("sendVerificationCode() - guard failed: mmId='\(uiState.mattermostId)', generation=\(uiState.selectedGeneration ?? -1)")
            return
        }

        joinLog("sendVerificationCode() - mmId='\(mmId)', generation=\(generation), name='\(uiState.name)'")

        Task {
            uiState.isSendingCode = true
            uiState.error = nil

            let result = await authRepository.requestSsafyVerification(
                mattermostId: mmId,
                generation: generation,
                name: uiState.name
            )

            switch result {
            case .success:
                joinLog("sendVerificationCode() SUCCESS")
                uiState.isSendingCode = false
                uiState.isCodeSent = true
            case .failure(let error):
                joinLog("sendVerificationCode() FAILED: \(error)")
                uiState.isSendingCode = false
                uiState.error = toJoinError(error)
            }
        }
    }

    func verifyCode() {
        let mmId = uiState.mattermostId.trimmingCharacters(in: .whitespacesAndNewlines)
        let code = uiState.verificationCode
        guard code.count == 6 else { return }

        Task {
            uiState.isVerifying = true
            uiState.error = nil

            let result = await authRepository.confirmSsafyVerification(mattermostId: mmId, code: code)

            switch result {
            case .success:
                uiState.isVerifying = false
                uiState.isVerified = true
                uiState.step = 3
            case .failure(let error):
                uiState.isVerifying = false
                uiState.error = toJoinError(error)
            }
        }
    }

    // MARK: - Step 3: Account Info

    func onEmailChange(_ value: String) {
        uiState.email = value
        uiState.error = nil
        uiState.isEmailAvailable = nil
    }

    func checkEmailDuplicate() {
        let email = uiState.email.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !email.isEmpty else { return }

        Task {
            uiState.isCheckingEmail = true
            uiState.error = nil

            let result = await authRepository.isEmailUsed(email: email)

            switch result {
            case .success(let isUsed):
                uiState.isCheckingEmail = false
                uiState.isEmailAvailable = !isUsed
            case .failure(let error):
                uiState.isCheckingEmail = false
                uiState.isEmailAvailable = nil
                uiState.error = toJoinError(error)
            }
        }
    }

    func onPasswordChange(_ value: String) {
        uiState.password = value
        uiState.error = nil
    }

    func onPasswordConfirmChange(_ value: String) {
        uiState.passwordConfirm = value
        uiState.error = nil
    }

    func signUp(onSuccess: @escaping () -> Void) {
        guard uiState.isStep3Valid else { return }
        guard let campus = uiState.selectedCampus,
              let generation = uiState.selectedGeneration,
              let ban = uiState.selectedClass else { return }

        Task {
            uiState.isLoading = true
            uiState.error = nil

            let info = SignUpInfo(
                email: uiState.email.trimmingCharacters(in: .whitespacesAndNewlines),
                password: uiState.password,
                name: uiState.name.trimmingCharacters(in: .whitespacesAndNewlines),
                studentId: uiState.studentId,
                campus: String(campus.id),
                generation: generation,
                clazz: ban.classNo ?? 0,
                mattermostId: uiState.mattermostId.trimmingCharacters(in: .whitespacesAndNewlines)
            )

            let result = await authRepository.signUp(info: info)

            switch result {
            case .success:
                uiState.isLoading = false
                uiState.isJoinSuccess = true
                onSuccess()
            case .failure(let error):
                uiState.isLoading = false
                uiState.error = toJoinError(error)
            }
        }
    }

    // MARK: - Helpers

    private func toJoinError(_ error: Error) -> JoinError {
        let message = error.localizedDescription
        if message.lowercased().contains("network") {
            return .network
        } else if message.contains("500") {
            return .server(message)
        } else {
            return .validation(message)
        }
    }

    func clearError() {
        uiState.error = nil
    }
}

File: android/ssabree/app/src/main/java/com/ssafy/ssabree/features/join/viewmodel/JoinViewModel.kt
package com.ssafy.ssabree.features.join.viewmodel

import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ssafy.ssabree.core.datasource.remote.model.Ban
import com.ssafy.ssabree.core.datasource.remote.model.Campus
import com.ssafy.ssabree.core.repository.AuthRepository
import com.ssafy.ssabree.core.repository.CampusRepository
import com.ssafy.ssabree.core.repository.model.SignUpInfo
import kotlinx.coroutines.launch
import java.net.SocketTimeoutException
import java.net.UnknownHostException

private const val TAG = "JoinViewModel"

sealed class JoinError(val message: String) {
    class Network(msg: String = "네트워크 연결을 확인해주세요.") : JoinError(msg)
    class Server(msg: String = "서버 오류가 발생했습니다.") : JoinError(msg)
    class Validation(msg: String) : JoinError(msg)
    class Unknown(msg: String = "알 수 없는 오류가 발생했습니다.") : JoinError(msg)
}

private fun Throwable.toJoinError(): JoinError = when {
    this is UnknownHostException -> JoinError.Network()
    this is SocketTimeoutException -> JoinError.Network()
    message?.contains("500") == true -> JoinError.Server()
    else -> JoinError.Unknown(message ?: "알 수 없는 오류")
}

private fun normalizeJoinValidationMessage(message: String): String {
    val lower = message.lowercase()
    return if (
        lower.contains("password") ||
        lower.contains("비밀번호") ||
        lower.contains("pw")
    ) {
        "비밀번호 형식이 올바르지 않습니다. 8자 이상, 영문/숫자/특수문자(@\$!%*#?&)를 포함해주세요."
    } else {
        message
    }
}

data class JoinUiState(
    // Step 1: 개인 정보
    val name: String = "",
    val studentId: String = "",

    // 기수 선택 (Step 1)
    val generations: List<Int> = listOf(14, 15),
    val selectedGeneration: Int? = null,
    val isGenerationDropdownExpanded: Boolean = false,

    // 캠퍼스 선택
    val campuses: List<Campus> = emptyList(),
    val selectedCampus: Campus? = null,
    val isCampusDropdownExpanded: Boolean = false,
    val isLoadingCampuses: Boolean = false,

    // 반 선택
    val campusClasses: List<Ban> = emptyList(),
    val classes: List<Ban> = emptyList(),
    val selectedClass: Ban? = null,
    val isClassDropdownExpanded: Boolean = false,
    val isLoadingClasses: Boolean = false,

    // Step 2: Mattermost 인증
    val mattermostId: String = "",
    val verificationCode: String = "",
    val isCodeSent: Boolean = false,
    val isVerified: Boolean = false,
    val isSendingCode: Boolean = false,
    val isVerifying: Boolean = false,

    // Step 3: 계정 정보
    val email: String = "",
    val isCheckingEmail: Boolean = false,
    val isEmailAvailable: Boolean? = null,  // null: 미확인 true: 사용 가능 false: 중복
    val password: String = "",
    val passwordConfirm: String = "",

    // 공통
    val step: Int = 1,
    val isLoading: Boolean = false,
    val error: JoinError? = null,
    val isJoinSuccess: Boolean = false,
) {
    val isStep1Valid: Boolean
        get() = name.isNotBlank() &&
            studentId.isNotBlank() &&
            selectedGeneration != null &&
            selectedCampus != null &&
            selectedClass != null

    val isStep2Valid: Boolean
        get() = verificationCode.length == 6

    val isStep3Valid: Boolean
        get() = email.isNotBlank() &&
                password.isNotBlank() &&
                passwordConfirm.isNotBlank() &&
                password == passwordConfirm &&
                !isLoading

    val passwordMismatch: Boolean
        get() = passwordConfirm.isNotBlank() && password != passwordConfirm
}

class JoinViewModel(
    private val authRepository: AuthRepository,
    private val campusRepository: CampusRepository
) : ViewModel() {

    var uiState by mutableStateOf(JoinUiState())
        private set

    init {
        loadCampuses()
    }

    private fun loadCampuses() {
        viewModelScope.launch {
            uiState = uiState.copy(isLoadingCampuses = true)

            campusRepository.getCampuses()
                .onSuccess { campusList ->
                    Log.d(TAG, "loadCampuses: succeed")
                    uiState = uiState.copy(
                        isLoadingCampuses = false,
                        campuses = campusList
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "loadCampuses: failed (${e.message})")
                    uiState = uiState.copy(
                        isLoadingCampuses = false,
                        error = e.toJoinError()
                    )
                }
        }
    }

    private fun loadClasses(campusId: Int) {
        viewModelScope.launch {
            uiState = uiState.copy(isLoadingClasses = true, classes = emptyList(), selectedClass = null)

            campusRepository.getClasses(campusId)
                .onSuccess { classList ->
                    Log.d(TAG, "loadClasses: succeed")
                    val filtered = filterClasses(classList, uiState.selectedGeneration)
                    uiState = uiState.copy(
                        isLoadingClasses = false,
                        campusClasses = classList,
                        classes = filtered,
                        selectedClass = null
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "loadClasses: failed (${e.message})")
                    uiState = uiState.copy(
                        isLoadingClasses = false,
                        error = e.toJoinError()
                    )
                }
        }
    }

    // --- Step Navigation ---

    fun goToStep(step: Int) {
        uiState = uiState.copy(step = step, error = null)
    }

    /**
     * 뒤로가기 처리
     * @return true: 이전 스텝으로 이동, false: 화면 종료 필요
     */
    fun goBack(): Boolean = when (uiState.step) {
        1 -> false  // Step 1: 화면 종료
        2 -> {
            uiState = uiState.copy(
                step = 1,
                error = null,
                mattermostId = "",
                verificationCode = "",
                isCodeSent = false,
                isVerified = false,
                isSendingCode = false,
                isVerifying = false,
                isGenerationDropdownExpanded = false
            )
            true
        }
        3 -> {
            uiState = uiState.copy(
                step = 2,
                error = null,
                mattermostId = "",
                verificationCode = "",
                isCodeSent = false,
                isVerified = false,
                isSendingCode = false,
                isVerifying = false,
                isGenerationDropdownExpanded = false,
                email = "",
                isCheckingEmail = false,
                isEmailAvailable = null,
                password = "",
                passwordConfirm = "",
                isLoading = false
            )
            true
        }
        else -> false
    }

    // --- Step 1 ---

    fun onNameChange(value: String) {
        uiState = uiState.copy(name = value)
    }

    fun onStudentIdChange(value: String) {
        uiState = uiState.copy(studentId = value.filter { it.isDigit() })
    }

    // --- 캠퍼스 선택 ---

    fun onCampusDropdownClick() {
        uiState = uiState.copy(isCampusDropdownExpanded = true)
    }

    fun onCampusDropdownDismiss() {
        uiState = uiState.copy(isCampusDropdownExpanded = false)
    }

    fun onCampusSelected(campus: Campus) {
        uiState = uiState.copy(
            selectedCampus = campus,
            isCampusDropdownExpanded = false,
            selectedClass = null,
            classes = emptyList(),
            campusClasses = emptyList()
        )
        loadClasses(campus.id)
    }

    // --- 반 선택 ---

    fun onClassDropdownClick() {
        if (uiState.selectedCampus != null && uiState.selectedGeneration != null) {
            uiState = uiState.copy(isClassDropdownExpanded = true)
        }
    }

    fun onClassDropdownDismiss() {
        uiState = uiState.copy(isClassDropdownExpanded = false)
    }

    fun onClassSelected(ban: Ban) {
        uiState = uiState.copy(
            selectedClass = ban,
            isClassDropdownExpanded = false
        )
    }

    private fun filterClasses(source: List<Ban>, generation: Int?): List<Ban> {
        if (generation == null) return emptyList()
        return source.filter { it.generation == generation }
    }

    // --- 기수 선택 ---

    fun onGenerationDropdownClick() {
        uiState = uiState.copy(isGenerationDropdownExpanded = true)
    }

    fun onGenerationDropdownDismiss() {
        uiState = uiState.copy(isGenerationDropdownExpanded = false)
    }

    fun onGenerationSelected(generation: Int) {
        val filteredClasses = filterClasses(uiState.campusClasses, generation)
        uiState = uiState.copy(
            selectedGeneration = generation,
            isGenerationDropdownExpanded = false,
            classes = filteredClasses,
            selectedClass = null
        )
    }

    // --- Step 2 ---

    fun onMattermostIdChange(value: String) {
        uiState = uiState.copy(mattermostId = value, error = null)
    }

    fun onVerificationCodeChange(value: String) {
        uiState = uiState.copy(
            verificationCode = value.filter { it.isDigit() }.take(6),
            error = null
        )
    }

    fun sendVerificationCode() {
        val mmId = uiState.mattermostId.trim()
        val generation = uiState.selectedGeneration
        val name = uiState.name
        if (mmId.isBlank() || generation == null) return

        viewModelScope.launch {
            uiState = uiState.copy(isSendingCode = true, error = null)

            authRepository.requestSsafyVerification(mmId, generation = generation, name = name)
                .onSuccess {
                    Log.d(TAG, "sendVerificationCode: succeed")
                    uiState = uiState.copy(
                        isSendingCode = false,
                        isCodeSent = true
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "sendVerificationCode: failed (${e.message})")
                    uiState = uiState.copy(
                        isSendingCode = false,
                        error = e.toJoinError()
                    )
                }
        }
    }

    fun verifyCode() {
        val mmId = uiState.mattermostId.trim()
        val code = uiState.verificationCode
        if (code.length != 6) return

        viewModelScope.launch {
            uiState = uiState.copy(isVerifying = true, error = null)

            authRepository.confirmSsafyVerification(mmId, code)
                .onSuccess {
                    Log.d(TAG, "verifyCode: succeed")
                    uiState = uiState.copy(
                        isVerifying = false,
                        isVerified = true,
                        step = 3
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "verifyCode: failed (${e.message})")
                    uiState = uiState.copy(
                        isVerifying = false,
                        error = e.toJoinError()
                    )
                }
        }
    }

    // --- Step 3 ---

    fun onEmailChange(value: String) {
        uiState = uiState.copy(email = value, error = null, isEmailAvailable = null)
    }

    fun checkEmailDuplicate() {
        val email = uiState.email.trim()
        if (email.isBlank()) return

        viewModelScope.launch {
            uiState = uiState.copy(isCheckingEmail = true, error = null)

            authRepository.isEmailUsed(email)
                .onSuccess { isUsed ->
                    Log.d(TAG, "checkEmailDuplicate: succeed (isUsed=$isUsed)")
                    uiState = uiState.copy(
                        isCheckingEmail = false,
                        isEmailAvailable = !isUsed,
                        error = null
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "checkEmailDuplicate: failed (${e.message})")
                    uiState = uiState.copy(
                        isCheckingEmail = false,
                        isEmailAvailable = null,
                        error = e.toJoinError()
                    )
                }
        }
    }

    fun onPasswordChange(value: String) {
        uiState = uiState.copy(password = value, error = null)
    }

    fun onPasswordConfirmChange(value: String) {
        uiState = uiState.copy(passwordConfirm = value, error = null)
    }

    fun signUp() {
        if (!uiState.isStep3Valid) return

        viewModelScope.launch {
            uiState = uiState.copy(isLoading = true, error = null)

            val info = SignUpInfo(
                email = uiState.email.trim(),
                password = uiState.password,
                name = uiState.name.trim(),
                studentNo = uiState.studentId.toIntOrNull(),
                campus = uiState.selectedCampus?.id,
                generation = uiState.selectedGeneration,
                classNo = uiState.selectedClass?.classNo,
                mattermostId = uiState.mattermostId.trim()
            )

            authRepository.signUp(info)
                .onSuccess {
                    Log.d(TAG, "signUp: succeed")
                    uiState = uiState.copy(
                        isLoading = false,
                        isJoinSuccess = true
                    )
                }
                .onFailure { e ->
                    Log.d(TAG, "signUp: failed (${e.message})")
                    val msg = e.message
                    uiState = uiState.copy(
                        isLoading = false,
                        error = if (msg.isNullOrBlank()) null else JoinError.Validation(normalizeJoinValidationMessage(msg))
                    )
                }
        }
    }
}
